feat: add Node.js RAG recipe

mhdawson opened the pull request.

mhdawson committed:
feat: add Node.js RAG recipe

Signed-off-by: Michael Dawson <mdawson@devrus.com>

mhdawson commented:
I've tested this with podman ai lab and all works well there.

I  also test running with `podman kube play build/rag-nodejs.yaml`.  That also works provided I set the chroma db container to ghcr.io/chroma-core/chroma:0.5.16 to match the version that podman ai desktop builds. 
* The chromadb container used based on the default in the shared Makefiles, comes from quay.io is much older (0.4.4) as it looks like the quay container for chromadb has not been updated for 8 months or so.    
* I think the old version of chromadb in the Makefiles is not an issue with this new recipe but something that should be addressed separately. Let me know if I should open an issue on the repo for that or if its a known issue.

mhdawson commented:
@Gregory-Pereira the second Node.js recipe, it should look familar as it reuses part of the Node.js chatbot recipe.

Gregory-Pereira commented on recipes/natural_language_processing/rag-nodejs/Makefile:
```diff
@@ -0,0 +1,67 @@
+RELATIVE_MODELS_PATH := ../../../models
+RELATIVE_TESTS_PATH := ../tests
+
+.PHONY: run-chromadb
+run: 
+	podman run -it -p $(CHROMADB_PORT):$(CHROMADB_PORT) -e CHROMADB_ENDPOINT=http://10.88.0.1:8000/v1 ${CHROMADB_IMAGE}
```
This stuff should be moved to the [chroma vectordb makefile](https://github.com/containers/ai-lab-recipes/blob/main/vector_dbs/chromadb/Makefile) and called via an override from here.

Gregory-Pereira commented on recipes/natural_language_processing/rag-nodejs/Makefile:
```diff
@@ -0,0 +1,67 @@
+	podman run -it -p $(CHROMADB_PORT):$(CHROMADB_PORT) -e CHROMADB_ENDPOINT=http://10.88.0.1:8000/v1 ${CHROMADB_IMAGE}
+
+# rag requires custom bootc because it uses an extra build-arg for CHROMADB_IMAGE (other apps use ../../common/Makefile.common target)
+.PHONY: bootc
+bootc: quadlet
+	"${CONTAINER_TOOL}"  build \
+	  $(ARCH:%=--arch %) \
+	  $(BUILD_ARG_FILE:%=--build-arg-file=%) \
+	  $(FROM:%=--from %) \
+	  $(AUTH_JSON:%=-v %:/run/containers/0/auth.json) \
+	  --security-opt label=disable \
+	  --cap-add SYS_ADMIN \
+	  --build-arg MODEL_IMAGE=$(MODEL_IMAGE) \
+	  --build-arg APP_IMAGE=$(APP_IMAGE) \
+	  --build-arg CHROMADB_IMAGE=$(CHROMADB_IMAGE) \
+	  --build-arg SERVER_IMAGE=$(SERVER_IMAGE) \
+	  --build-arg "SSHPUBKEY=$(SSH_PUBKEY)" \
+	  -f bootc/$(CONTAINERFILE) \
+	  -t ${BOOTC_IMAGE} .
+	@echo ""
+	@echo "Successfully built bootc image '${BOOTC_IMAGE}'."
+	@echo "You may now convert the image into a disk image via bootc-image-builder"
+	@echo "or the Podman Desktop Bootc Extension.  For more information, please refer to"
+	@echo "   * https://github.com/osbuild/bootc-image-builder"
+	@echo "   * https://github.com/containers/podman-desktop-extension-bootc"
```
I would rather see this update the [common makefile to allow an extra bootc buildarg](https://github.com/containers/ai-lab-recipes/blob/main/recipes/common/Makefile.common#L85-L105), which would reduce code duplicaiton. What im imagining is some variable like `EXTRA_BOOTC_BUILD_ARGS`:
Makefile.common:
```Makefile.common
EXTRA_BOOTC_BUILD_ARGS ?=
...
.PHONY: bootc
bootc: quadlet growfs
	podman build \
	  $(ARCH:%=--arch %) \
	  $(BUILD_ARG_FILE:%=--build-arg-file=%) \
	  $(FROM:%=--from %) \
	  $(AUTH_JSON:%=-v %:/run/containers/0/auth.json) \
	  --security-opt label=disable \
	  --cap-add SYS_ADMIN \
	  --build-arg MODEL_IMAGE=$(MODEL_IMAGE) \
	  --build-arg APP_IMAGE=$(APP_IMAGE) \
	  --build-arg SERVER_IMAGE=$(SERVER_IMAGE) \
	  --build-arg "SSHPUBKEY=$(SSH_PUBKEY)" \
      ${EXTRA_BOOTC_BUILD_ARGS} \
	  -f bootc/$(CONTAINERFILE) \
	  -t ${BOOTC_IMAGE} bootc
	@echo ""
	@echo "Successfully built bootc image '${BOOTC_IMAGE}'."
	@echo "You may now convert the image into a disk image via bootc-image-builder"
	@echo "or the Podman Desktop Bootc Extension.  For more information, please refer to"
	@echo "   * https://github.com/osbuild/bootc-image-builder"
	@echo "   * https://github.com/containers/podman-desktop-extension-bootc"
	
```

Then from the recipe you would call it like: `EXTRA_BOOTC_BUILD_ARGS="--build-arg something=something_else" make bootc`, etc.

Gregory-Pereira replied:
Seems like this is copied from the other rag example, however we should have extended the base in other cases. Since I dont want to set  a new precendent, we will consider these a non issue for now and refactor.

Gregory-Pereira commented on recipes/natural_language_processing/rag-nodejs/Makefile:
```diff
@@ -0,0 +1,67 @@
+
+# rag requires custom quadlet target for CHROMADB_IMAGE substitution
+# (other apps use ../../common/Makefile.common target)
+.PHONY: quadlet
+quadlet:
+	# Modify quadlet files to match the server, model and app image
+	mkdir -p build
+	sed -e "s|SERVER_IMAGE|${SERVER_IMAGE}|" \
+	    -e "s|APP_IMAGE|${APP_IMAGE}|g" \
+	    -e "s|MODEL_IMAGE|${MODEL_IMAGE}|g" \
+	    -e "s|CHROMADB_IMAGE|${CHROMADB_IMAGE}|g" \
+	    -e "s|APP|${APP}|g" \
+	    quadlet/${APP}.image \
+	    > build/${APP}.image
+	sed -e "s|SERVER_IMAGE|${SERVER_IMAGE}|" \
+	    -e "s|APP_IMAGE|${APP_IMAGE}|g" \
+	    -e "s|MODEL_IMAGE|${MODEL_IMAGE}|g" \
+	    -e "s|CHROMADB_IMAGE|${CHROMADB_IMAGE}|g" \
+	    quadlet/${APP}.yaml \
+	    > build/${APP}.yaml
+	cp quadlet/${APP}.kube build/${APP}.kube
```
Same as above, lets find a non-invasive way to extend the common option to suit your needs here. This is an important pattern because it helps reduce complexity when debugging, leaving 1 section of code to traverse and step through rather than 2.

Gregory-Pereira commented on recipes/natural_language_processing/rag-nodejs/Makefile:
```diff
@@ -0,0 +1,67 @@
+
+# rag requires custom bootc-run because it uses an extra port for chromadb
+# (other apps use ../../common/Makefile.common target)
+.PHONY: bootc-run
+bootc-run:
+	podman run -d --rm --name $(APP)-bootc -p 8080:8501 -p 8090:8000 --privileged \
+	  $(AUTH_JSON:%=-v %:/run/containers/0/auth.json) \
+	  $(BOOTC_IMAGE) /sbin/init
```
Same as above

Gregory-Pereira commented on recipes/natural_language_processing/rag-nodejs/bootc/Containerfile:
```diff
@@ -0,0 +1,48 @@
+ARG MODEL_IMAGE=quay.io/ai-lab/granite-7b-lab:latest
+ARG APP_IMAGE=quay.io/ai-lab/${RECIPE}:latest
+ARG SERVER_IMAGE=quay.io/ai-lab/llamacpp_python:latest
+ARG CHROMADBImage=quay.io/ai-lab/chromadb
```
I dont believe this is correct, should be: 
```suggestion
ARG CHROMADB_IMAGE=quay.io/ai-lab/chromadb:latest
```

mhdawson replied:
Its a direct copy from the existing rag recipe - https://github.com/containers/ai-lab-recipes/blob/7fbdf03a8dadbf5a41da283fdc1c07ea8d1ec61a/recipes/natural_language_processing/rag/bootc/Containerfile#L21

mhdawson replied:
Updated

Gregory-Pereira reviewed (commented):
Overall great PR. I just had some process points. If there is a reason to rush this in like a blog before EOY or some other time dependent deliverable we could postpone those to a follow up PR, but if not Id prefer we implement through common makefile.

mhdawson commented:
@Gregory-Pereira in terms of the suggested changes, those are all to files which are almost a direct copy of those being used for the existing rag recipe. Is that something that should be part of this PR or a cleanup that covers both?   I have enough understanding to copy/tweak those files, but would a bigger lift to understand well enough to refactor.

I just copied them over and updated to use `rag-nodesjs` instead of `rag` for the most part.

There is no major rush, but would like to avoid having other PRs having me to rework.

mhdawson commented:
I'm off on PTO until Dec 23 so if I don't respond to comments/questions that's why. I'll check in when I return on the 23rd.

mhdawson commented:
I'm back for the next few days but then out again until Jan 2nd. In any case I expect lots of people will be out and will check in again when I'm back in Jan.

mhdawson committed:
squash: address comments

Signed-off-by: Michael Dawson <mdawson@devrus.com>

mhdawson committed:
squash: pin chromadb ver to match current client

pin the chromadb version when using quadlet and bootc to the
same one used when run with podman ai lab. Chromadb seems to
break comapibility regularly and the client must be compatible
with the chromadb version used.

Signed-off-by: Michael Dawson <mdawson@devrus.com>

mhdawson commented:
@Gregory-Pereira I pushed a commit with the changes you requested.

I then pushed another commit which set the chromadb container used by quadlet and bootc to match what is used when the recipe is used with the podman ai lab extension. It seems that chromadb is still changing quickly and the client version seems to need to match the chromadb  version in use.  This addresses the issue mentioned eariler and I think is better than changing the common chromadb version that is used for quadlet in the python recipe as it might have an older client  and that might explain why an older container is being use.

After these changes the I test that the recipe still works with the podman desktop ai extension as well as when you follow the instructions to run it with quadlet.

Gregory-Pereira reviewed (approved):
/lgtm

mhdawson commented:
@Gregory-Pereira many thanks for your help.
